hazelcast/src/main/java/com/hazelcast/internal/util/ConcurrencyUtil.java

```diff
@@ -45,14 +46,15 @@ public String toString() {
         }
     };
 
-    // Default executor for async callbacks: ForkJoinPool.commonPool() or a thread-per-task executor when
-    // the common pool does not support parallelism
+    // Default executor for async callbacks: Executors.newFixedThreadPool of ForkJoinPool.commonPool()
+    // size or a thread-per-task executor when the common pool does not support parallelism
     private static Executor defaultAsyncExecutor;
 
     static {
         Executor asyncExecutor;
-        if (ForkJoinPool.getCommonPoolParallelism() > 1) {
-            asyncExecutor = ForkJoinPool.commonPool();
+        int commonPoolParallelism = ForkJoinPool.getCommonPoolParallelism();
+        if (commonPoolParallelism > 1) {
+            asyncExecutor = Executors.newFixedThreadPool(commonPoolParallelism);
```

Comment:
This change is breaking the contract of our async methods eg see [`IMap` javadoc](https://github.com/hazelcast/hazelcast/blob/14ef0dbefa4a6fa00a50a8873aeea02ec6c3f5e4/hazelcast/src/main/java/com/hazelcast/map/IMap.java#L210-L222)

Reply:
I agree, but is that contract a "must-have"? cannot we change it? I cannot imagine situation where user may want to depend on the thread the Async action is invoked on. 

By introducing this change (yes, breaking change, I agree) we no longer rely on FJP.commonPool, so it's way less possible someone will "starve" our actions by adding some blocking to common pool. If we can change the contract, I think we should.